### PR TITLE
Python3-six-removal

### PR DESCRIPTION
**Pull request by a-detiste**
I never got my device to work ... sadness,
I'm better at software that hardware.

I gifted identical device to a hackerlab and while googling for support I found this one back https://iooner.io/epson-vfd-hack/ :-) :-) 

So anyway the weird mixture of Six & Typing ticked me and there I can help.

https://wiki.debian.org/Python3-six-removal

But ... ! While removing six I got scared and now I'd prefer to cleanup the annotations first.

(I can also package this for Debian)

### CODE REVIEW

**Comment by a-detiste** [comment]
Ok there's something wrong with the CI itself

File: src/escpos/escpos.py
Comment (patkan) [inline src/escpos/escpos.py:1437]:
```suggestion
            self.printer.text(f"{line}\n")
```

File: src/escpos/escpos.py
Comment (patkan) [inline src/escpos/escpos.py:267]:
```suggestion
            xm = b"\x01" if high_density_horizontal else b"\x02"
```
Shouldn't that be hex? (Otherwise I guess you are using the byte representation of ASCII "1"

File: src/escpos/printer/network.py
Comment (patkan) [inline src/escpos/printer/network.py:119]: What is the motivation behind this assertion? If I remember correctly, the device property should open the connection automatically on access.

File: src/escpos/printer/network.py
Comment (patkan) [inline src/escpos/printer/network.py:124]: Same here as in line 119

**Review by patkan — COMMENTED** [review]
Thank you for the contribution.
First review pass, I will check again with more time. Please have a look at the comments.
As far as I know the type annotation for a None return type is technically not necessary, but doesn't do any harm either, so why not.

It would be great if you would package for Debian, maybe we should discuss in a discussion what would be needed in order to do it from this repository (on a separate branch?)

File: src/escpos/escpos.py
Comment (a-detiste) [reply src/escpos/escpos.py:1437]: I'm not used to these new kind of f-string.

File: src/escpos/escpos.py
Comment (a-detiste) [reply src/escpos/escpos.py:267]: WHOOPS !

File: src/escpos/printer/network.py
Comment (a-detiste) [reply src/escpos/printer/network.py:119]: I use `mypy --strict`to lint my code; mypy can't know that self.device has been initialized somehow in the lifecycle of the object, it's the only way to tell it `yes it's fine`

File: src/escpos/escpos.py
Comment (patkan) [reply src/escpos/escpos.py:1437]: In the end it is just a more compact form of the format syntax (without the boilerplate like this: `"{thing}".format(thing=thing)`)

File: src/escpos/printer/network.py
Comment (patkan) [reply src/escpos/printer/network.py:119]: OK, agreed, that is a bit much to expect from a static code checker :-)

**Comment by a-detiste** [comment]
> what would be needed in order to do it from this repository (on a separate branch?)

this is against the common practices ... this way you would get a .deb file not a Debian package.

Debian is still obsessed about .tar.gz and these would then be loaded on https://salsa.debian.org/
(to create another .tar.gz for the build servers :-) )

**Comment by patkan** [comment]
> > what would be needed in order to do it from this repository (on a separate branch?)
> 
> this is against the common practices ... this way you would get a .deb file not a Debian package.
> 
> Debian is still obsessed about .tar.gz and these would then be loaded on https://salsa.debian.org/ (to create another .tar.gz for the build servers :-) )

OK, we can do whatever is best practice. The only thing that would be important to me is that we find a way that could be "recoverable" if you would stop packaging python-escpos. I have tried (with the organization) to have the python-escpos package at least somehow hosted in a such a way that somebody else could step up and take this over with (hopefully) not too much trouble, should I stop doing this some day. (Ideally, there would be a transition, but you never know)

**Comment by a-detiste** [comment]
I'm aware of this too strong ownership of packages in the Debian project.

I got shouted at by some survivalist living of the grid 11 months in a row
and going online once a year. "It's my package"

I will package it under the "Python team" with a disclaimer that anyone else can update it.

**Comment by patkan** [comment]
> I'm aware of this too strong ownership of packages in the Debian project.
> 
> I got shouted at by some survivalist living of the grid 11 months in a row and going online once a year. "It's my package"
> 
> I will package it under the "Python team" with a disclaimer that anyone else can update it.

That would be great :-)



I have created a PR that refactors everything to f-strings: #608. If you do not mind merging or rebasing your branch, I would merge that into master.

**Comment by a-detiste** [comment]
>  If you do not mind merging

I do not use an ide ... still nano ... :-|

**Comment by patkan** [comment]
> > If you do not mind merging
> 
> I do not use an ide ... still nano ... :-|

I have checked and it is rather easy to merge. If that is OK for you, I would merge #608 and merge the changes into your branch, then you just have to pull :-)

**Comment by a-detiste** [comment]
Yes please.

**Comment by patkan** [comment]
> Yes please.

Done

**Comment by patkan** [comment]
Now there is only one thing left in the CI:

```
  py38: commands[0]> pytest
  ImportError while loading conftest '/home/runner/work/python-escpos/python-escpos/test/conftest.py'.
  test/conftest.py:4: in <module>
      from escpos.printer import LP, CupsPrinter, Dummy, File, Network, Serial, Usb, Win32Raw
  .tox/py38/lib/python3.8/site-packages/escpos/printer/__init__.py:4: in <module>
      from .cups import CupsPrinter
  .tox/py38/lib/python3.8/site-packages/escpos/printer/cups.py:16: in <module>
      from ..escpos import Escpos
  .tox/py38/lib/python3.8/site-packages/escpos/escpos.py:1363: in <module>
      class EscposIO:
  .tox/py38/lib/python3.8/site-packages/escpos/escpos.py:1441: in EscposIO
      self, type: type[BaseException], value: BaseException, traceback: TracebackType
  E   TypeError: 'type' object is not subscriptable
```

I will soon be offline for today and cannot approve the CI runs. But you can also execute the tests locally with e.g. `tox -e py310` if you want to try out changes.

**Comment by patkan** [comment]
I think the issue was that one: 369e709fc29d82b0d7d92009dc9c6f37c1190e65, feel free to revert if that is not what you meant.

**Comment by a-detiste** [comment]
I would have proposed a `from __future__ import annotations`

**Comment by a-detiste** [comment]
Your solution is the perfect one for python3.8, which I'm not used to work with (I do 2.6, 2.7, 3.7, 3.11, 3.12)

**Comment by patkan** [comment]
> Your solution is the perfect one for python3.8, which I'm not used to work with (I do 2.6, 2.7, 3.7, 3.11, 3.12)

OK, but then I would prefer that we git revert 369e709fc29d82b0d7d92009dc9c6f37c1190e65, because I would prefer more the way how it will be done in the future.

**Comment by patkan** [comment]
> > Your solution is the perfect one for python3.8, which I'm not used to work with (I do 2.6, 2.7, 3.7, 3.11, 3.12)
> 
> OK, but then I would prefer that we git revert [369e709](https://github.com/python-escpos/python-escpos/commit/369e709fc29d82b0d7d92009dc9c6f37c1190e65), because I would prefer more the way how it will be done in the future.

I have cleaned up my commit

**Comment by a-detiste** [comment]
> I would prefer more the way how it will be done in the future.

pyupgrade can do it automaticaly for you when you bump your baseline.

`Optional[a]` becomes `a | None`
`Union[a|b]` becomes `a | b`

most uppercase imports from `typing` are replaced by lowercase ones from standard library
`List` -> `list`

but it's easier to keep it 3.8 coherant for now.

**Comment by a-detiste** [comment]
> src/escpos/escpos.py:1278: error: Incompatible return value type (got "bytes", expected "list[int]")  [return-value]

this looks like a real bug

**Comment by patkan** [comment]
> > src/escpos/escpos.py:1278: error: Incompatible return value type (got "bytes", expected "list[int]")  [return-value]
> 
> this looks like a real bug

Is this about the return value of the cups driver? If yes, maybe @belono can help? It looks like the return value of _read is not compatible with the rest of the lib for this driver.

**Comment by a-detiste** [comment]
It could works thanks to duck typing: in some context bytes can be itterrated as single int s. But it would mire readable to stick with one data type for this object interface.

**Comment by belono** [comment]
> > > src/escpos/escpos.py:1278: error: Incompatible return value type (got "bytes", expected "list[int]")  [return-value]
> > 
> > 
> > this looks like a real bug
> 
> Is this about the return value of the cups driver? If yes, maybe @belono can help? It looks like the return value of _read is not compatible with the rest of the lib for this driver.

Should the cups driver's `_read()` method return bytes? If so, would this return values be correct?
`b"0"` for the 'idle' state
`b"8"` for 'printing a job' or 'stopped' states

**Comment by a-detiste** [comment]
If it's really about bytes emulated/encapsulated by cups it's better to use bytes all the way.

**Comment by a-detiste** [comment]
Ok I'm further.

`error: "Escpos" has no attribute "close" ` -> indeed the destructor will call in cascade a method that doesn't exist

**Comment by patkan** [comment]
> Ok I'm further.
> 
> `error: "Escpos" has no attribute "close" ` -> indeed the destructor will call in cascade a method that doesn't exist

The method should be implemented by the printer drivers. Is this not the case?

**Comment by a-detiste** [comment]
There's Escpos && EscposIO:

One close() is missing

```
class Escpos

    def __del__(self):
        """Call self.close upon deletion."""
        self.close()


class EscposIO:
    r"""ESC/POS Printer IO object.
       def __init__(
        self, printer: Escpos, autocut: bool = True, autoclose: bool = True, **kwargs
    ) -> None:
        """Initialize object.
        :param printer: An EscPos-printer object
        """
        self.printer = printer

 def close(self) -> None:
        """Close printer.

        Called upon closing the `with`-statement.
        """
        self.printer.close()
```

**Comment by a-detiste** [comment]
It's a baseclass, so maybe just addig "def close(): pass" is enough

**Comment by a-detiste** [comment]
I know nothing about win32, but the typing in the stubs _feels_ wrong.

```
/usr/lib/python3/dist-packages/win32-stubs/win32print.pyi:def WritePrinter(hprinter: _win32typing.PyPrinterHANDLE, buf: str): ...
```

**Comment by a-detiste** [comment]
BTW the `) -> None:` has one specific purpose: it tells mypy that annotation job has been done and it can proceed with further checking. This makes more hints show up.

**Comment by patkan** [comment]
Hi @a-detiste, are you ready with the change? (CI is passing)
Would it be possible to change the mypy ci step to be more strict? (Or do we have remaining issues there?)

**Comment by a-detiste** [comment]
Hi, I would like to slow download before my Holydays (from 20), so let's wrap this up.

mypy --strict generates 400 errors, it's wat too early.

**Comment by patkan** [comment]
Ok, well this would be quite a lot still, haven't checked mypy directly.

Should we merge this pr now? Then you can open a new one for the next steps when you get to it

**Comment by a-detiste** [comment]
Yes please merge.

I will track on my side the dependecies that needs:
- annotations
- packaging in Debian

**Comment by a-detiste** [comment]
I need to package python3-barcode first

**Comment by a-detiste** [comment]
please include examples/ in mypy run.
this helped me understanding & fixing this previous typing error

-tox.ini:commands = mypy src test
+tox.ini:commands = mypy src test examples

```
-    def __init__(
-        idVendor: str = "",
-        idProduct: str = "",
-        usb_args: Dict[str, str] = {},
    
+    def __init__(
+        self,
+        idVendor: Optional[int] = None,
+     idProduct: Optional[int] = None,
+     usb_args: Dict[str, Union[str, int]] = {},
```

**Review by patkan — APPROVED** [review]
For me these changes would be ok. @belono what do you think?

**Comment by belono** [comment]
I left a comment in the [cups.py](https://github.com/python-escpos/python-escpos/pull/607/files#diff-b7fcda07446fce62bb555abb5722fbc061848fdf0fbc2744bd38e122d4935ee4R208) and another one in [test_printer_cups.py](https://github.com/python-escpos/python-escpos/pull/607/files/45829b8ecb9d6ec2f33336da5b572801a49be233..254eae07ef345bef788e108176b659c2353e8449#diff-8fba37f83986b38b6d5a02c153258f0511aff1e2163c72663397c2b983686e5aR176) some days ago. :smile: 

Please, fix that before merge.

**Comment by patkan** [comment]
> I left a comment in the [cups.py](https://github.com/python-escpos/python-escpos/pull/607/files#diff-b7fcda07446fce62bb555abb5722fbc061848fdf0fbc2744bd38e122d4935ee4R208) and another one in [test_printer_cups.py](https://github.com/python-escpos/python-escpos/pull/607/files/45829b8ecb9d6ec2f33336da5b572801a49be233..254eae07ef345bef788e108176b659c2353e8449#diff-8fba37f83986b38b6d5a02c153258f0511aff1e2163c72663397c2b983686e5aR176) some days ago. 😄
> 
> Please, fix that before merge.

Are you sure that you have sent the review? Could it be still pending? I cannot see those comments unfortunately. :confused:

**Review by belono — CHANGES_REQUESTED** [review]
Please, modify the `CupsPrinter._read()` method and its test.

File: src/escpos/printer/cups.py
Comment (a-detiste) [reply src/escpos/printer/cups.py:208]: I think a lot of constants.py should be rewriten using Enum / IntEnum instead of having magic numbers that needs commenting here and there... later

Comment (patkan) [reply src/escpos/printer/cups.py:208]: you could use `RT_MASK_ONLINE` in this case.

Comment (belono) [reply src/escpos/printer/cups.py:208]: The reason I didn't suggest RT_MASK_ONLINE is that its name is not consistent with the meaning of the returning value. According to [Esc/Pos](https://download4.epson.biz/sec_pubs/pos/reference_en/escpos/dle_eot.html), the int value 8 is for **Offline** status.

We could modify that constant and add a new one for offline:
```python
RT_MASK_ONLINE: int = 0
RT_MASK_OFFLINE: int = 8
```
Then we could do:
```python
if not state or state in [4, 5]:
    return RT_MASK_OFFLINE  # offline
return RT_MASK_ONLINE  # online
```

Comment (patkan) [reply src/escpos/printer/cups.py]: I have created an issue so that we do not forget and introduced your first example, @belono so that we have a solution for now.

**Pull request merged**
